File: js/tests/parsing.test.ts
/** 어댑터 정규화 검증 - 실측 골든 픽스처 재생 (Kotlin/Python 과 동일 정답지). */
import assert from "node:assert/strict";
import { test } from "node:test";
import { AuthError, Decimal, KisClient, KiwoomClient, NextClient, isOpenStatus, krxTickRound } from "../src/index.js";

const stub = (responses: Record<string, unknown>[]) => {
  const queue = [...responses];
  return async () => queue.shift()!;
};

test("next: quote 파싱 (v1.3 — outcome=OK 만, 등락률 %→비율, KST 시각)", async () => {
  const client = new NextClient("k", "s");
  (client as any).request = stub([{
    quotes: [
      { symbol: "AAPL", outcome: "OK", session: "REGULAR", requestedAt: "2026-09-10T23:10:00+09:00",
        price: "308.91", previousClose: "333.43", change: "-24.52", changeRate: "-7.3539",
        bidPrice: null, bidSize: null, askPrice: null, askSize: null,
        volume: "132756799", lastTradeAt: "2026-09-10T23:09:58+09:00" },
      { symbol: "NOPE", outcome: "NOT_FOUND", session: "CLOSED", requestedAt: "2026-09-10T23:10:00+09:00" },
    ],
  }]);
  const quotes = await client.getQuotes(["AAPL", "NOPE"]);
  assert.deepEqual(quotes.map((q) => q.symbol), ["AAPL"]);
  const [q] = quotes;
  assert.equal(q.price.toString(), "308.91");
  assert.equal(q.bidPrice, null);
  assert.equal(q.changeRate!.toString(), "-0.073539");
  assert.equal(q.volume, 132756799);
  assert.equal(q.timestamp.toISOString(), "2026-09-10T14:09:58.000Z");
});

test("next: 캔들 time 에 오프셋이 없으면 KST 로 해석", async () => {
  const client = new NextClient("k", "s");
  (client as any).request = stub([{
    symbol: "AAPL", interval: "1d", nextCursor: null,
    candles: [{ time: "2026-09-09T22:30:00", session: "REGULAR", open: "339.73", high: "344.56",
                low: "337.35", close: "338.19", volume: "56298904" }],
  }]);
  const [c] = await client.getCandles("AAPL", "1d", 3);
  assert.equal(c.timestamp.toISOString(), "2026-09-09T13:30:00.000Z");
  assert.equal(c.volume, 56298904);
});

test("next: 캘린더 KST 세션 → 뉴욕 현지 HH:mm, status → open", async () => {
  const client = new NextClient("k", "s");
  (client as any).request = stub([{
    calendar: [
      { date: "2026-09-10", status: "OPEN", holidayName: null, sessions: [
        { type: "PRE", open: "2026-09-10T17:00:00+09:00", close: "2026-09-10T22:30:00+09:00" },
        { type: "REGULAR", open: "2026-09-10T22:30:00+09:00", close: "2026-09-11T05:00:00+09:00" }] },
      { date: "2026-11-27", status: "HALF_DAY", holidayName: "Day After Thanksgiving", sessions: [
        { type: "REGULAR", open: "2026-11-27T23:30:00+09:00", close: "2026-11-28T03:00:00+09:00" }] },
      { date: "2026-12-25", status: "CLOSED", holidayName: "Christmas", sessions: [] },
    ],
  }]);
  const days = await client.getCalendar();
  assert.equal(days[0].open, true);
  assert.equal(days[0].timezone, "America/New_York");
  assert.deepEqual(days[0].regular, { start: "09:30", end: "16:00" });
  assert.equal(days[1].regular!.end, "13:00"); // 반장일 (EST)
  assert.equal(days[1].holiday, "Day After Thanksgiving");
  assert.equal(days[2].open, false);
  assert.equal(days[2].regular, null);
});

test("next: 계좌 총평가 = cashAmount + 보유 평가금액, 손익률 %→비율", async () => {
  const client = new NextClient("k", "s");
  (client as any).request = stub([
    { accountId: "acc_main", currency: "USD", cashAmount: "1000.50", requestedAt: "2026-09-10T23:10:00+09:00" },
    { currency: "USD", holdings: [
      { symbol: "AAPL", name: "Apple Inc.", quantity: "2", sellableQuantity: "2", averageBuyPrice: "300.00",
        currentPrice: "310.00", purchaseAmount: "600.00", evaluationAmount: "620.00",
        evaluationPnl: "20.00", evaluationPnlRate: "3.3333" }], requestedAt: "2026-09-10T23:10:00+09:00" },
  ]);
  const account = await client.getAccount();
  assert.equal(account.cash.toString(), "1000.5");
  assert.equal(account.portfolioValue.toString(), "1620.5");
  (client as any).request = stub([
    { currency: "USD", holdings: [
      { symbol: "AAPL", name: "Apple Inc.", quantity: "2", averageBuyPrice: "300.00", evaluationAmount: "620.00",
        evaluationPnl: "20.00", evaluationPnlRate: "3.3333" }] },
  ]);
  const [h] = await client.getHoldings();
  assert.equal(h.avgEntryPrice.toString(), "300");
  assert.equal(h.marketValue!.toString(), "620");
  assert.equal(h.unrealizedPnlRate!.toString(), "0.033333");
});

test("next: 주문 상세 requestId → clientOrderId, 생성 본문에 market·clientOrderId", async () => {
  const client = new NextClient("k", "s");
  (client as any).request = stub([{
    orderId: "ord_1", requestId: "s-1", market: "US", symbol: "AAPL", side: "BUY", orderType: "LIMIT",
    timeInForce: "DAY", status: "PARTIALLY_FILLED", quantity: "10", filledQuantity: "4", limitPrice: "200",
    avgFillPrice: "199.5", requestedAt: "2026-09-10T23:10:00+09:00", updatedAt: "2026-09-10T23:12:00+09:00",
  }]);
  const order = await client.getOrder("ord_1");
  assert.equal(order.clientOrderId, "s-1");
  assert.equal(order.status, "PARTIALLY_FILLED");
  assert.equal(order.orderType, "LIMIT");
  assert.equal(order.submittedAt!.toISOString(), "2026-09-10T14:10:00.000Z");

  const sent: Record<string, unknown>[] = [];
  (client as any).request = async (_m: string, _p: string, opts: { json?: Record<string, unknown> }) => {
    sent.push(opts.json!);
    return { orderId: "ord_2", market: "US", status: "SUBMITTED", requestedAt: "2026-09-10T23:10:00+09:00" };
  };
  await client.createOrder({ symbol: "AAPL", side: "BUY", orderType: "LIMIT", quantity: new Decimal(1),
                             limitPrice: new Decimal(200), clientOrderId: "s-2" });
  assert.equal(sent[0].market, "US");
  assert.equal(sent[0].clientOrderId, "s-2");
  assert.equal(sent[0].limitPrice, "200");
});

test("next: PENDING_CANCEL 은 open 으로 분류 (v1.3 부록 D)", async () => {
  const client = new NextClient("k", "s");
  (client as any).request = stub([{
    orderId: "ord_1", status: "PENDING_CANCEL", symbol: "AAPL", side: "BUY",
    orderType: "LIMIT", quantity: "1", limitPrice: "200", filledQuantity: "0",
  }]);
  const order = await client.cancelOrder("ord_1");
  assert.equal(order.status, "PENDING_CANCEL");
  assert.equal(isOpenStatus(order.status), true);
});

test("next: v1.3 공통 헤더 (X-Request-Id, X-Next-Account-Id)", async () => {
  const seen: { url: string; headers: Record<string, string> }[] = [];
  const restore = (globalThis as any).fetch;
  (globalThis as any).fetch = async (url: string, init: { headers: Record<string, string> }) => {
    seen.push({ url, headers: init.headers });
    const body = url.endsWith("/v1/oauth/token")
      ? { access_token: "tok", token_type: "Bearer", expires_in: 43200 }
      : url.endsWith("/v1/account/holdings")
        ? { currency: "USD", holdings: [] }
        : { accountId: "acc_main", currency: "USD", cashAmount: "1" };
    return { status: 200, text: async () => JSON.stringify(body) };
  };
  try {
    const client = new NextClient("k", "s", "acc_main", "http://next.test");
    await client.getAccount();
  } finally {
    (globalThis as any).fetch = restore;
  }
  const [, account] = seen;
  assert.equal(account.headers["X-Next-Account-Id"], "acc_main");
  assert.equal(account.headers["X-Nextsecurities-Account"], undefined);
  assert.match(account.headers["X-Request-Id"], /^[A-Za-z0-9._-]{1,64}$/);
});

test("next: 토큰 발급 실패는 OAuth 표준 에러 형식", async () => {
  const restore = (globalThis as any).fetch;
  (globalThis as any).fetch = async () => ({
    status: 401, text: async () => JSON.stringify({ error: "invalid_client", error_description: "인증 실패" }),
  });
  try {
    const client = new NextClient("k", "s", "acc_main", "http://next.test");
    await assert.rejects(client.getQuotes(["AAPL"]), (e: any) =>
      e instanceof AuthError && e.errorCode === "invalid_client" && String(e.message).includes("인증 실패"));
  } finally {
    (globalThis as any).fetch = restore;
  }
});

test("kis: %단위 등락률 -> 비율", async () => {
  const client = new KisClient("k", "s", "50199202");
  (client as any).call = stub([{
    output: { stck_prpr: "239500", acml_vol: "27393580", prdy_vrss: "-23000", prdy_ctrt: "-8.76" },
  }]);
  const [q] = await client.getQuotes(["005930"]);
  assert.equal(q.price.toString(), "239500");
  assert.equal(q.changeRate!.toString(), "-0.0876");
});

test("kis: 일봉 최신순 -> 과거→최신 정렬", async () => {
  const client = new KisClient("k", "s", "50199202");
  (client as any).call = stub([{
    output2: [
      { stck_bsop_date: "20260731", stck_oprc: "304", stck_hgpr: "310", stck_lwpr: "300", stck_clpr: "308", acml_vol: "1" },
      { stck_bsop_date: "20260730", stck_oprc: "333", stck_hgpr: "334", stck_lwpr: "329", stck_clpr: "333", acml_vol: "2" },
    ],
  }]);
  const candles = await client.getCandles("005930", "1d", 2);
  assert.ok(candles[0].timestamp < candles[1].timestamp);
  assert.equal(candles[1].close.toString(), "308");
});

test("kis: 추적 주문의 체결을 보유수량 변화로 판정", async () => {
  const client = new KisClient("k", "s", "50199202");
  let qty = new Decimal(2);
  (client as any).getHoldings = async () => [{ symbol: "005930", quantity: qty }];
  (client as any).call = async () => ({ output: { ODNO: "0000035787" } });

  const order = await client.createOrder({
    symbol: "005930", side: "BUY", orderType: "LIMIT",
    quantity: new Decimal(1), limitPrice: new Decimal(185400),
  });
  assert.equal((await client.getOrder(order.orderId)).status, "SUBMITTED");
  qty = new Decimal(3); // 체결로 보유 증가
  assert.equal((await client.getOrder(order.orderId)).status, "FILLED");
});

test("kiwoom: 부호 접두/zero-padded 정규화", async () => {
  const client = new KiwoomClient("k", "s");
  (client as any).call = stub([
    { cur_prc: "-239500", trde_qty: "27393580", pred_pre: "-23000", flu_rt: "-8.76" },
  ]);
  const [q] = await client.getQuotes(["005930"]);
  assert.equal(q.price.toString(), "239500");     // 부호 제거
  assert.equal(q.change!.toString(), "-23000");   // 대비는 부호 유지
  assert.equal(q.changeRate!.toString(), "-0.0876");
});

test("kiwoom: 계좌 zero-padded 금액", async () => {
  const client = new KiwoomClient("k", "s");
  (client as any).call = stub([
    { entr: "000000100000000", ord_alow_amt: "000000100000000" },
    { prsm_dpst_aset_amt: "000000100000000", tot_evlt_amt: "000000000000000" },
  ]);
  const account = await client.getAccount();
  assert.equal(account.cash.toString(), "100000000");
});

test("kiwoom: 보유 A 프리픽스 제거", async () => {
  const client = new KiwoomClient("k", "s");
  (client as any).balance = stub([{
    acnt_evlt_remn_indv_tot: [
      { stk_cd: "A005930", rmnd_qty: "000000000001", pur_pric: "000000239500",
        cur_prc: "-240000", evlt_amt: "000000240000", evltv_prft: "500", prft_rt: "0.21" },
    ],
  }]);
  const [h] = await client.getHoldings();
  assert.equal(h.symbol, "005930");
  assert.equal(h.currentPrice!.toString(), "240000");
});

test("krx 호가단위 보정", () => {
  assert.equal(krxTickRound(new Decimal("246258.99")).toString(), "246000"); // 20만~50만: 500원
  assert.equal(krxTickRound(new Decimal("197650")).toString(), "197600");    // 5만~20만: 100원
  assert.equal(krxTickRound(new Decimal("1999")).toString(), "1999");        // ~2천: 1원
});

File: go/parsing_test.go
package hermetix

// 어댑터 정규화 검증 - 실측 골든 픽스처 재생 (Kotlin/Python/JS 와 동일 정답지).

import (
	"testing"
	"time"

	"github.com/shopspring/decimal"
)

func stubCalls(responses []map[string]any) func() map[string]any {
	i := 0
	return func() map[string]any {
		r := responses[i]
		i++
		return r
	}
}

// v1.3: outcome=OK 만 공통 모델로, 등락률 % → 비율, KST 시각 파싱
func TestNextQuoteParsing(t *testing.T) {
	c := NewNextClient("k", "s")
	c.call = func(method, path string, account bool, jsonBody map[string]any) (map[string]any, error) {
		return map[string]any{"quotes": []any{
			map[string]any{
				"symbol": "AAPL", "outcome": "OK", "session": "REGULAR", "requestedAt": "2026-09-10T23:10:00+09:00",
				"price": "308.91", "previousClose": "333.43", "change": "-24.52", "changeRate": "-7.3539",
				"bidPrice": nil, "askPrice": nil, "volume": "132756799", "lastTradeAt": "2026-09-10T23:09:58+09:00",
			},
			map[string]any{"symbol": "NOPE", "outcome": "NOT_FOUND", "session": "CLOSED", "requestedAt": "2026-09-10T23:10:00+09:00"},
		}}, nil
	}
	quotes, err := c.GetQuotes([]string{"AAPL", "NOPE"})
	if err != nil {
		t.Fatal(err)
	}
	if len(quotes) != 1 || quotes[0].Symbol != "AAPL" {
		t.Fatalf("NOT_FOUND 종목은 제외돼야 한다: %+v", quotes)
	}
	if quotes[0].Price.String() != "308.91" {
		t.Fatalf("price = %s", quotes[0].Price)
	}
	if quotes[0].BidPrice != nil {
		t.Fatal("bidPrice should be nil")
	}
	if quotes[0].ChangeRate.String() != "-0.073539" {
		t.Fatalf("changeRate = %s", quotes[0].ChangeRate)
	}
	if quotes[0].Volume != 132756799 {
		t.Fatalf("volume = %d", quotes[0].Volume)
	}
	if got := quotes[0].Timestamp.UTC().Format(time.RFC3339); got != "2026-09-10T14:09:58Z" {
		t.Fatalf("timestamp = %s", got)
	}
}

func TestNextCandleTimeWithoutOffsetIsKST(t *testing.T) {
	c := NewNextClient("k", "s")
	c.call = func(method, path string, account bool, jsonBody map[string]any) (map[string]any, error) {
		return map[string]any{"symbol": "AAPL", "interval": "1d", "candles": []any{map[string]any{
			"time": "2026-09-09T22:30:00", "session": "REGULAR", "open": "339.73", "high": "344.56",
			"low": "337.35", "close": "338.19", "volume": "56298904",
		}}}, nil
	}
	candles, err := c.GetCandles("AAPL", Day1, 3)
	if err != nil {
		t.Fatal(err)
	}
	if got := candles[0].Timestamp.UTC().Format(time.RFC3339); got != "2026-09-09T13:30:00Z" {
		t.Fatalf("timestamp = %s", got)
	}
}

func TestNextCalendarKSTToNewYork(t *testing.T) {
	c := NewNextClient("k", "s")
	c.call = func(method, path string, account bool, jsonBody map[string]any) (map[string]any, error) {
		return map[string]any{"calendar": []any{
			map[string]any{"date": "2026-09-10", "status": "OPEN", "holidayName": nil, "sessions": []any{
				map[string]any{"type": "PRE", "open": "2026-09-10T17:00:00+09:00", "close": "2026-09-10T22:30:00+09:00"},
				map[string]any{"type": "REGULAR", "open": "2026-09-10T22:30:00+09:00", "close": "2026-09-11T05:00:00+09:00"},
			}},
			map[string]any{"date": "2026-11-27", "status": "HALF_DAY", "holidayName": "Day After Thanksgiving", "sessions": []any{
				map[string]any{"type": "REGULAR", "open": "2026-11-27T23:30:00+09:00", "close": "2026-11-28T03:00:00+09:00"},
			}},
			map[string]any{"date": "2026-12-25", "status": "CLOSED", "holidayName": "Christmas", "sessions": []any{}},
		}}, nil
	}
	days, err := c.GetCalendar()
	if err != nil {
		t.Fatal(err)
	}
	if !days[0].Open || days[0].Timezone != "America/New_York" || days[0].Regular.Start != "09:30" || days[0].Regular.End != "16:00" {
		t.Fatalf("day0 = %+v regular=%+v", days[0], days[0].Regular)
	}
	if !days[1].Open || days[1].Regular.End != "13:00" || days[1].Holiday != "Day After Thanksgiving" {
		t.Fatalf("half day = %+v regular=%+v", days[1], days[1].Regular)
	}
	if days[2].Open || days[2].Regular != nil {
		t.Fatalf("closed day = %+v", days[2])
	}
}

func TestNextAccountPortfolioValue(t *testing.T) {
	c := NewNextClient("k", "s")
	c.call = func(method, path string, account bool, jsonBody map[string]any) (map[string]any, error) {
		if path == "/v1/account/holdings" {
			return map[string]any{"currency": "USD", "holdings": []any{map[string]any{
				"symbol": "AAPL", "name": "Apple Inc.", "quantity": "2", "sellableQuantity": "2", "averageBuyPrice": "300.00",
				"currentPrice": "310.00", "purchaseAmount": "600.00", "evaluationAmount": "620.00",
				"evaluationPnl": "20.00", "evaluationPnlRate": "3.3333",
			}}}, nil
		}
		return map[string]any{"accountId": "acc_main", "currency": "USD", "cashAmount": "1000.50", "requestedAt": "2026-09-10T23:10:00+09:00"}, nil
	}
	account, err := c.GetAccount()
	if err != nil {
		t.Fatal(err)
	}
	if account.Cash.String() != "1000.5" || account.PortfolioValue.String() != "1620.5" {
		t.Fatalf("account = %+v", account)
	}
	holdings, _ := c.GetHoldings()
	if holdings[0].AvgEntryPrice.String() != "300" || holdings[0].MarketValue.String() != "620" || holdings[0].UnrealizedPnlRate.String() != "0.033333" {
		t.Fatalf("holding = %+v", holdings[0])
	}
}

func TestNextOrderDetailAndCreateBody(t *testing.T) {
	c := NewNextClient("k", "s")
	var sent map[string]any
	c.call = func(method, path string, account bool, jsonBody map[string]any) (map[string]any, error) {
		if method == "POST" {
			sent = jsonBody
			return map[string]any{"orderId": "ord_2", "market": "US", "status": "SUBMITTED", "requestedAt": "2026-09-10T23:10:00+09:00"}, nil
		}
		return map[string]any{
			"orderId": "ord_1", "requestId": "s-1", "market": "US", "symbol": "AAPL", "side": "BUY", "orderType": "LIMIT",
			"timeInForce": "DAY", "status": "PENDING_CANCEL", "quantity": "10", "filledQuantity": "4", "limitPrice": "200",
			"requestedAt": "2026-09-10T23:10:00+09:00", "updatedAt": "2026-09-10T23:12:00+09:00",
		}, nil
	}
	order, err := c.GetOrder("ord_1")
	if err != nil {
		t.Fatal(err)
	}
	if order.ClientOrderID != "s-1" || order.Status != PendingCancel || !order.Status.IsOpen() || order.OrderType != Limit {
		t.Fatalf("order = %+v", order)
	}
	if got := order.SubmittedAt.UTC().Format(time.RFC3339); got != "2026-09-10T14:10:00Z" {
		t.Fatalf("submittedAt = %s", got)
	}
	limit := decimal.NewFromInt(200)
	if _, err := c.CreateOrder(CreateOrderRequest{Symbol: "AAPL", Side: Buy, OrderType: Limit, Quantity: decimal.NewFromInt(1), LimitPrice: &limit, ClientOrderID: "s-2"}); err != nil {
		t.Fatal(err)
	}
	if sent["market"] != "US" || sent["clientOrderId"] != "s-2" || sent["limitPrice"] != "200" || sent["orderType"] != "LIMIT" {
		t.Fatalf("body = %+v", sent)
	}
}

func TestKisQuotePercentToRate(t *testing.T) {
	c := NewKisClient("k", "s", "50199202")
	c.call = func(method, path, trID string, query, jsonBody map[string]string) (map[string]any, error) {
		return map[string]any{"output": map[string]any{
			"stck_prpr": "239500", "acml_vol": "27393580", "prdy_vrss": "-23000", "prdy_ctrt": "-8.76",
		}}, nil
	}
	quotes, err := c.GetQuotes([]string{"005930"})
	if err != nil {
		t.Fatal(err)
	}
	if quotes[0].Price.String() != "239500" {
		t.Fatalf("price = %s", quotes[0].Price)
	}
	if quotes[0].ChangeRate.String() != "-0.0876" {
		t.Fatalf("changeRate = %s", quotes[0].ChangeRate)
	}
}

func TestKisTrackedOrderFillByHoldings(t *testing.T) {
	c := NewKisClient("k", "s", "50199202")
	holdingQty := decimal.NewFromInt(2)
	c.call = func(method, path, trID string, query, jsonBody map[string]string) (map[string]any, error) {
		if trID == "VTTC8434R" { // 잔고
			return map[string]any{
				"output1": []any{map[string]any{"pdno": "005930", "hldg_qty": holdingQty.String(), "pchs_avg_pric": "0"}},
				"output2": []any{map[string]any{"dnca_tot_amt": "0", "tot_evlu_amt": "0"}},
			}, nil
		}
		return map[string]any{"output": map[string]any{"ODNO": "0000035787"}}, nil
	}

	limitPrice := decimal.NewFromInt(185400)
	order, err := c.CreateOrder(CreateOrderRequest{
		Symbol: "005930", Side: Buy, OrderType: Limit,
		Quantity: decimal.NewFromInt(1), LimitPrice: &limitPrice,
	})
	if err != nil {
		t.Fatal(err)
	}
	got, _ := c.GetOrder(order.OrderID)
	if got.Status != Submitted {
		t.Fatalf("status = %s", got.Status)
	}
	holdingQty = decimal.NewFromInt(3) // 체결로 보유 증가
	got, _ = c.GetOrder(order.OrderID)
	if got.Status != Filled {
		t.Fatalf("status = %s, want FILLED", got.Status)
	}
}

func TestKiwoomSignAndPadding(t *testing.T) {
	c := NewKiwoomClient("k", "s")
	c.call = func(path, apiID string, jsonBody map[string]string) (map[string]any, error) {
		return map[string]any{
			"cur_prc": "-239500", "trde_qty": "27393580", "pred_pre": "-23000", "flu_rt": "-8.76",
		}, nil
	}
	quotes, err := c.GetQuotes([]string{"005930"})
	if err != nil {
		t.Fatal(err)
	}
	if quotes[0].Price.String() != "239500" { // 부호 제거
		t.Fatalf("price = %s", quotes[0].Price)
	}
	if quotes[0].Change.String() != "-23000" { // 대비는 부호 유지
		t.Fatalf("change = %s", quotes[0].Change)
	}
}

func TestKiwoomHoldingsStripAPrefix(t *testing.T) {
	c := NewKiwoomClient("k", "s")
	c.call = func(path, apiID string, jsonBody map[string]string) (map[string]any, error) {
		return map[string]any{"acnt_evlt_remn_indv_tot": []any{map[string]any{
			"stk_cd": "A005930", "rmnd_qty": "000000000001", "pur_pric": "000000239500",
			"cur_prc": "-240000", "evlt_amt": "000000240000", "evltv_prft": "500", "prft_rt": "0.21",
		}}}, nil
	}
	holdings, err := c.GetHoldings()
	if err != nil {
		t.Fatal(err)
	}
	if holdings[0].Symbol != "005930" {
		t.Fatalf("symbol = %s", holdings[0].Symbol)
	}
	if holdings[0].CurrentPrice.String() != "240000" {
		t.Fatalf("currentPrice = %s", holdings[0].CurrentPrice)
	}
	if holdings[0].AvgEntryPrice.String() != "239500" { // zero-padded
		t.Fatalf("avgEntryPrice = %s", holdings[0].AvgEntryPrice)
	}
}

func TestKrxTickRound(t *testing.T) {
	cases := map[string]string{
		"246258.99": "246000", // 20만~50만: 500원
		"197650":    "197600", // 5만~20만: 100원
		"1999":      "1999",   // ~2천: 1원
	}
	for input, want := range cases {
		price, _ := decimal.NewFromString(input)
		if got := KrxTickRound(price).String(); got != want {
			t.Fatalf("KrxTickRound(%s) = %s, want %s", input, got, want)
		}
	}
}

// ------------------------------------------------------------- 0.6.0: 환경 · 심볼 접두

func TestNextEnvironmentMustMatchKeyPrefix(t *testing.T) {
	if err := NewNextClient("pk_test_x", "s").SetEnvironment(Live); err == nil {
		t.Fatal("pk_test_ 키로 LIVE 는 에러")
	}
	if err := NewNextClient("pk_live_x", "s").SetEnvironment(Paper); err == nil {
		t.Fatal("pk_live_ 키로 PAPER 는 에러")
	}
	c := NewNextClient("pk_live_x", "s")
	if err := c.SetEnvironment(Live); err != nil || c.Environment() != Live {
		t.Fatalf("LIVE 설정 실패: %v", err)
	}
	if NewNextClient("k", "s").Environment() != Paper {
		t.Fatal("기본은 PAPER")
	}
}

func TestNextMarketPrefixedSymbol(t *testing.T) {
	c := NewNextClient("k", "s")
	seen := ""
	c.call = func(method, path string, account bool, jsonBody map[string]any) (map[string]any, error) {
		seen = path
		return map[string]any{"quotes": []any{map[string]any{"symbol": "AAPL", "outcome": "OK", "price": "1", "requestedAt": "2026-09-10T23:10:00+09:00"}}}, nil
	}
	quotes, err := c.GetQuotes([]string{"US:AAPL"})
	if err != nil || quotes[0].Symbol != "US:AAPL" || seen != "/v1/market/quotes?symbols=AAPL" {
		t.Fatalf("quotes=%+v path=%s err=%v", quotes, seen, err)
	}
	if _, err := c.GetQuotes([]string{"KRX:005930"}); err == nil {
		t.Fatal("미지원 시장은 에러")
	}
}

func TestKisKiwoomEnvironmentHostAndTR(t *testing.T) {
	paper := NewKisClient("k", "s", "50199202")
	live := NewKisClient("k", "s", "50199202").SetEnvironment(Live)
	if paper.BaseURL() != KisPaperURL || paper.tr("TTC0802U") != "VTTC0802U" {
		t.Fatalf("paper = %s %s", paper.BaseURL(), paper.tr("TTC0802U"))
	}
	if live.BaseURL() != KisLiveURL || live.tr("TTC0802U") != "TTTC0802U" || live.Environment() != Live {
		t.Fatalf("live = %s %s", live.BaseURL(), live.tr("TTC0802U"))
	}
	if custom := NewKisClient("k", "s", "c").SetBaseURL("http://custom").SetEnvironment(Live); custom.BaseURL() != "http://custom" {
		t.Fatal("직접 지정한 호스트는 유지")
	}
	if NewKiwoomClient("k", "s").SetEnvironment(Live).BaseURL() != KiwoomLiveURL || NewKiwoomClient("k", "s").BaseURL() != KiwoomPaperURL {
		t.Fatal("kiwoom 호스트")
	}
}
